aspnetcore/mvc/models/file-uploads.md

```diff
@@ -28,13 +28,13 @@ Use caution when providing users with the ability to upload files to a server. A
 
 Security steps that reduce the likelihood of a successful attack are:
 
-* Upload files to a dedicated file upload area on the system, preferably to a non-system drive. Using a dedicated location makes it easier to impose security restrictions on uploaded files. Disable execute permissions on the file upload location.&dagger;
-* Never persist uploaded files in the same directory tree as the app.&dagger;
-* Use a safe file name determined by the app. Don't use a file name provided by the user or the untrusted file name of the uploaded file.&dagger; To display an untrusted file name in a UI or in a logging message, HTML-encode the value.
-* Only allow a specific set of approved file extensions.&dagger;
-* Check the file format signature to prevent a user from uploading a masqueraded file.&dagger; For example, don't permit a user to upload an *.exe* file with a *.txt* extension.
-* Verify that client-side checks are also performed on the server.&dagger; Client-side checks are easy to circumvent.
-* Check the size of an uploaded file and prevent uploads that are larger than expected.&dagger;
+* Upload files to a dedicated file upload area, preferably to a non-system drive. A dedicated location makes it easier to impose security restrictions on uploaded files. Disable execute permissions on the file upload location.&dagger;
+* Do **not** persist uploaded files in the same directory tree as the app.&dagger;
+* Use a safe file name determined by the app. Don't use a file name provided by the user or the untrusted file name of the uploaded file.&dagger; HTML encode the untrusted file name when displaying it. For example, logging the file name or displaying in UI (Razor automatically HTML encodes output).
+* Allow only approved file extensions for the app's design specification.&dagger;
+* Check the file format signature to prevent a user from uploading a masqueraded file.&dagger; For example, don't permit uploading an *.exe* file with a *.txt* extension.
```

Comment:
Not quite sure how to do this. Is there a reliable way to tell if somebody's trying to smuggle an executable file with a text extension? Should the server forbid uploading the all-text [EICAR test file](https://en.wikipedia.org/wiki/EICAR_test_file)?

If we can't show how to reliably do something I'd err on the side of not including it in a guideline.

Reply:
The example here uses the first few bytes. See the sample app ...

https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/models/file-uploads/samples/3.x/SampleApp/Utilities/FileHelpers.cs#L18-L51

It's an approach recommended by one of the Azure engineers ...

https://docs.microsoft.com/en-us/azure/security/develop/threat-modeling-tool-input-validation#controls-users

Reply:
I'd be curious to know a little more about what the author of that approach/code thinks.

Ping @msmbaldwin and @barbkess ... you are GH contributors to the following topic:

https://docs.microsoft.com/en-us/azure/security/develop/threat-modeling-tool-input-validation#controls-users

Do you know the source of the approach/code in the *Ensure appropriate controls are in place when accepting files from users* section?